### PR TITLE
Annotations - Added parsing of IRT, RT, State and StateModel

### LINKED ISSUE #10677: Contents of annotations not (completely) visible
I have not used any external library, i have used acrobat reader to annotate, insert and delete. I had taken the screenshot form your demo link only. Here by i will attach my pdf file as well as screenshot of demo link which you have given in the GitHub.

[Passasjeren_strukturert (4) 11.pdf](https://github.com/mozilla/pdf.js/files/3001745/Passasjeren_strukturert.4.11.pdf)

Screen Shot 2019-03-25 at 11 24 38 AM
Screen Shot 2019-03-25 at 11 24 48 AM

_Originally posted by @rakeshkasinathan in https://github.com/mozilla/pdf.js/issues/10672#issuecomment-476064327_

### PR DESCRIPTION
**Pull request by vlastimilmaca**
Hi, this is a last piece of #10716 .

This PR adds parsing of IRT, RT and IT into MarkupAnnotation. 
If RT is Group, it takes all properties from IRT annotation and sets them into current annotation, which is what should happen according to spec (see page 393, in the middle). This also happens for popups, which should fix #10677, where popup on strikeout annotation should have correct 'Contents' text. 

Parsing of contents has been moved from Markup annotation into Annotation, because according to spec, 'Contents' is common to all Annotations (see page 383, table 184).

IT is also parsed, but I'm not sure about it's meaning or usability in UI, but its easy to parse, so it is there …

There is also added 'isMarkup' property into Markup annotation data set, which should help in building new UI features using RT/IRT (discussion threads, states etc). 
ie. if isMarkup == true, it should be displayed in comments sidebar (if thats gonna be a thing), otherwise it should be skipped for that part of UI.

Finally there is parsing of State and StateModel in TextAnnotation, which is also preparation for new UI features, but it's not gonna impact anything for now.

Future annotation UI update should probably solve these two main issues:
- displaying replies inside of popup of IRT annotation or inside of other UI element (maybe new sidebar, or new "extended" popup for IRT annotation)
- not rendering annotations with RT=R or StateModel into canvas as they should be represented inside their IRT annotation's UI element.

Btw. As I was going through spec again, I noticed, that 'CreationDate' is a thing only for Markup Annotation, so maybe it should be moved too…

Thank you for reviews :).

### CODE REVIEW

**Comment by vlastimilmaca** [comment]
I'll fix this after #10833 is merged, so it is clean. Please try to look in to this if there is anything else I should fix. I'm quite short on free time these days ...

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: If I read the specification correctly, this shouldn't be necessary for the IRT/RT logic, so if that's the case please remove it from this patch to keep the functionality limited to IRT/RT. If need be, it can later be added in a follow-up patch.

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: This needs to be indented with two extra spaces.

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: Also here, please remove any properties that weren't in the original code and aren't used anywhere yet to keep the patch small. The same applies to the `else` branch below for `subject`.

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: You can remove this comment since that's already clear from the code itself.

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: Please remove this newline.

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: The spec states: "The state is not specified in the annotation itself but in a separate text annotation that refers to the original annotation by means of its IRT (“in reply to”) entry."

Should we check if an `IRT` entry is available here? Moreover, should this text annotation be rendered (which I think is the case at the moment) or should those annotations not be rendered because they only contain the state of another annotation? I think it's rather odd that the spec uses a different annotation for keeping state of another annotation, let alone it being a `Text` annotation, but that seems to be how it works...

Whether or not an annotation should be rendered is controlled in https://github.com/mozilla/pdf.js/blob/master/src/display/annotation_layer.js#L358-L359, so you could add a check for `parameters.data.state !== null` there.

Finally, I think the data attributes for an annotation type should always be provided and set to `null` if they are not used. This way the API is consistent: you always get all attributes on which you can do a check. In this case, the `state` and `stateModel` properties would alway be provided, but set to `null` if they don't have a value.

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: Let's simplify this to just `if (parentItem.get('RT').name === AnnotationReplyType.GROUP) {` because `RT` must always exist because of the `has` check above.

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: Let's add a comment after the `if` statement: "Subordinate annotations in a group should inherit the group attributes from the primary annotation." (taken from the spec)

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: Let's replace this comment with: "Subordinate annotations in a group should inherit the group attributes from the primary annotation." (taken from the spec)

File: src/shared/util.js
Comment (timvandermeij) [inline src/shared/util.js]: Just for readability, could you put these new entries in the correct (alphabetical) order?

**Review by timvandermeij — CHANGES_REQUESTED** [review]
Looks good! If you rebase onto the current master and incorporate these comments, you can simplify the patch even further. I only haven't looked at the tests yet because I think they will also change a bit after the rebase and comments, but I will look at those when that is done. Thank you!

File: src/core/annotation.js
Comment (Snuffleupagus) [reply src/core/annotation.js]: *I'm going to comment on this __once__ here, but this problem exists in many places.*

Just because a particular object exists in a `Dict`, that does *not* imply that it's actually of the desired type (i.e. `Name` here). Hence you need to check that as well, before attempting to access a potentially non-existent `name` property, since there will potentially be errors thrown otherwise.

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js]: Similar to my other comment, you need to actually validate this data before calling `toString()` on it to avoid errors.

File: src/core/annotation.js
Comment (vlastimilmaca) [reply src/core/annotation.js]: Yeah this is tricky … Maybe, as there is viewable and printable in Annotation base class, there should also exist renderable, which will tell, wheter annotation should be rendered into canvas ([added into opList (see code)](https://github.com/mozilla/pdf.js/blob/master/src/core/document.js#L243) ). Viewable will then tell, wheter it can be passed out of worker at all ([added into getAnnotationsData(see code)](https://github.com/mozilla/pdf.js/blob/master/src/core/document.js#L299) response). 

annotation.viewable is now used in both of those cases, but IMO it should be splitted. So in core/document.js there should exist isAnnotationRenderable as is now (which will check for annotation.renderable - for opList) and isAnnotationViewable (which will check for annotation.viewable - for getAnnotationsData). 

RT=R and State annotations would be renderable=false, but viewable=true.  Standard annotations and Group annotations (probably?) would be true,true. And those, which are now viewable=false would be false,false. But I would rather keep that into some future PRs as it might break a thing or two ...

File: src/core/annotation.js
Comment (vlastimilmaca) [inline src/core/annotation.js]: I'm not sure, if this check is OK.

**Comment by vlastimilmaca** [comment]
Thank you for checking this! I'm sorry that I didn't find time earlier as you probably already forgot what this PR was about :( (I almost did). I hope I addressed all requests, but I'm not sure if I correctly handled @Snuffleupagus request for checking getRaw() return values. If you have an idea on how to make it better, please let me know. Thanks !

PS. I'll squash it when it's ok.

File: src/core/annotation.js
Comment (vlastimilmaca) [reply src/core/annotation.js]: Removed.

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js]: What type is this supposed to be?

If `Ref`, then you need to check `isRef(...)` below; similar for other types of course.

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js]: This looks way too complicated, why can't you simply replace all of this block with *just* the following instead?
```javascript
const rt = dict.get('RT');
this.data.replyType = isName(rt) ? rt.name : AnnotationReplyType.REPLY;
```

File: src/core/annotation.js
Comment (Snuffleupagus) [reply src/core/annotation.js]: Same as above.

Comment (Snuffleupagus) [reply src/core/annotation.js]: See comment above.

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js]: This can be made a lot simpler, if you look at the [definition of the `isName` helper function](https://github.com/mozilla/pdf.js/blob/d0892e46e2b0d16ecd7f720f36f36202ae1c12cd/src/core/primitives.js#L284-L286):
```javascript
if (isName(parentItem.get('RT'), AnnotationReplyType.GROUP)) {
```

File: src/core/annotation.js
Comment (vlastimilmaca) [reply src/core/annotation.js]: Honestly, I'm not sure … it's the same thing as parent in popup, so I used same approach as the one that is already in master ([https://github.com/mozilla/pdf.js/blob/master/src/core/annotation.js#L989](https://github.com/mozilla/pdf.js/blob/master/src/core/annotation.js#L989)). But it is probably Ref … so I do it like this? 
isRef(dict.get("IRT")) ? dict.getRaw("IRT").toString() : null

File: src/core/annotation.js
Comment (Snuffleupagus) [inline src/core/annotation.js]: This can be made a lot simpler, and the `.has(...)` check is unnecessary here as well, if you look at the [definition of the `isName` helper function](https://github.com/mozilla/pdf.js/blob/d0892e46e2b0d16ecd7f720f36f36202ae1c12cd/src/core/primitives.js#L284-L286):
```javascript
const rt = parentItem.get('RT');
if (isName(rt, AnnotationReplyType.GROUP)) {
  ...
}
```

File: src/core/annotation.js
Comment (vlastimilmaca) [reply src/core/annotation.js]: I was not sure what get('RT') would do, if 'RT' is not there, but if that's ok, than it can be as you suggested :). Will do.

File: src/core/annotation.js
Comment (vlastimilmaca) [reply src/core/annotation.js]: That's cool, thank you :) ... didn't cross my mind that it can have two arguments.

File: src/core/annotation.js
Comment (Snuffleupagus) [reply src/core/annotation.js]: Please check the code, and you'll see what it returns in that case: https://github.com/mozilla/pdf.js/blob/d0892e46e2b0d16ecd7f720f36f36202ae1c12cd/src/core/primitives.js#L83-L97

File: src/core/annotation.js
Comment (Snuffleupagus) [reply src/core/annotation.js]: > isRef(dict.get("IRT")) ? dict.getRaw("IRT").toString() : null

That won't work at all since `dict.get` dereferences `Ref` objects, whereas `dict.getRaw` doesn't as its name suggests.
```javascript
const rawIRT = dict.getRaw('IRT');
this.data.inReplyTo = isRef(rawIRT) ? rawIRT.toString() : null;
```

File: src/shared/util.js
Comment (Snuffleupagus) [inline src/shared/util.js]: Any particular reason for not making all of these added structures enums?

Comment (vlastimilmaca) [reply src/shared/util.js]: They are stored as strings in annotations, which is what I need to compare it against. Javascript can do string enums?

File: src/core/annotation.js
Comment (vlastimilmaca) [reply src/core/annotation.js]: Yeah … thats gonna be undefined … so !isName … so it is ok :).

File: src/core/annotation.js
Comment (vlastimilmaca) [reply src/core/annotation.js]: I'm sorry ... I'm really tired ... after I looked into get method you showed me, it hit me :). Thanks.

File: src/shared/util.js
Comment (vlastimilmaca) [reply src/shared/util.js]: Should I also order it alphabetically, or keep it together in 'logic' order like this? Move it somewhere ?

File: src/core/annotation.js
Comment (vlastimilmaca) [reply src/core/annotation.js]: You are right. Removed.

File: src/core/annotation.js
Comment (vlastimilmaca) [reply src/core/annotation.js]: Removed.

File: src/core/annotation.js
Comment (timvandermeij) [reply src/core/annotation.js]: That sounds like a good approach, but I agree it's best to keep that for a follow-up PR given its complexity.

File: src/shared/util.js
Comment (timvandermeij) [reply src/shared/util.js]: I don't think the order matters much here, but it's usally nice to keep the `export` list at the bottom ordered so we can quickly spot if something is exported or not.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: This unit test can be removed now as it was included in the PR you rebased upon.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: The indentation is a bit off here and in other parts of the added code in this file. This and the code below is part of the `describe` block and should therefore be indented with two extra spaces.

**Comment by vlastimilmaca** [comment]
Round #2 of fixes. Thank you very much for reviews, it is very educational for me :) … I'm sorry you have to go through it though :(.

**Comment by vlastimilmaca** [comment]
Just rebase...

**Comment by timvandermeij** [comment]
I noticed that you also fixed the indentation in the unit tests; great! I'm hoping to play around with this and take another look very soon, but the changes that have been made so far look really good.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: In my previous PR I already added a `describe` block for this (see https://github.com/mozilla/pdf.js/blob/master/test/unit/annotation_spec.js#L320), so you can simply place your new tests in there so we don't have to create the block here again.

Comment (vlastimilmaca) [reply test/unit/annotation_spec.js]: Didn't see that :). Thank you. Moved.

**Comment by vlastimilmaca** [comment]
Should be fine, don't know why it won't run lint test :/

**Comment by timvandermeij** [comment]
Hm, me neither, but it isn't caused by your patch. Most likely a new intermittent failure that we haven't seen before. Let's try again...

/botio-windows fonttest

**Comment by vlastimilmaca** [comment]
well I did a force-push and then after few minutes, probably before it get to finish lint test, I did another force push … that may broke it? Should I try another force push?  But I don't know what to fix … it's perfect now :D.

**Comment by vlastimilmaca** [comment]
So …  what about this ? :)

**Comment by timvandermeij** [comment]
I have finally worked my way through the backlog of pull requests, releases and other tasks, so I'll soon have time for this pull request again. Sorry for the delay!

**Comment by vlastimilmaca** [comment]
Nice :), no problem. We should probably start by waking up Travis somehow … can you run some force command ?

**Comment by timvandermeij** [comment]
I triggered Travis by closing and reopening the pull request.

File: src/core/annotation.js
Comment (timvandermeij) [inline src/core/annotation.js]: Please revert this change to keep the diff small.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: Change this to `should not parse state model and state when not defined`.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: Could you place this `describe` block below the `AnnotationBorderStyle` `describe` block for a more logical ordering?

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: Remove the newline here, just like above, for consistency (applies to other blocks in tests below too since the newlines are a bit inconsistent throughout these changes).

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: Change this to `should parse state model and state when defined`.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: Formatting of these blocks should be like this in terms of indentation:
https://github.com/mozilla/pdf.js/blob/6e0e725af401cb3a7a96aba15624b1c625967cc6/test/unit/annotation_spec.js#L113-L117
(applies to other blocks in tests below too)

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: I would suggest to leave out this line and the one below it because this test only tests `stateModel` and `state` being set correctly. The other fields should be covered in other tests, which you already have below, to keep the test's scope limited.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: Could you move this last test to the `PopupAnnotation`'s `describe` block at:
https://github.com/mozilla/pdf.js/blob/6e0e725af401cb3a7a96aba15624b1c625967cc6/test/unit/annotation_spec.js#L1674

That's a better place for this test since it's relevant in that scope.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: Let's make this test only about not parsing IRT/RT, so change this to `should not parse IRT/RT when not defined` and remove the `T` test since that is not really related to IRT/RT not being parsed.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: Change this to `should parse IRT and set default RT when not defined`.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: There are basically two situations when IRT/RT is parsed: either it's a group annotation (and its contents should be inherited) or it's a reply annotation. How about updating this test and the one below so that this is clear from the name and so all fields that can be changed, namely title, contents, creation date, modification date, `hasPopup` and color, are checked. That way it's clear which two scenarios we expect and we make sure that all fields that can be changed are in fact tested instead of only the `Contents` field. Both tests are already pretty close to this, but needs some tweaks to get there, mainly by changing the name and testing all fields.

This one can then be called `should parse IRT/RT for a group type` and the other one can then be called `should parse IRT/RT for a reply type`.

**Review by timvandermeij — COMMENTED** [review]
You'll see that the remaining comments are mostly about the tests because I hadn't looked into those in detail before. The non-test code looks really good this way. I think that we can merge this pull request once these comments are taken care of. I was quite busy last time and finally found the time to properly review this, so sorry for the delay and thank you for your patience!

**Review by vlastimilmaca — COMMENTED** [review]
I agree with everything suggested. Just had one question about repositioning `describe` block.

File: test/unit/annotation_spec.js
Comment (vlastimilmaca) [reply test/unit/annotation_spec.js]: You're right, it is weird this way … so I tried to follow this pattern (I found it in other tests in this file and I think it is more readable)

```
Dict1

Dict2

Ref1
Ref2
xRef

ResultCheck
```

In one occasion, where I needed Ref1 before Dict2 I went with this pattern
```
Ref1
Dict1

Ref2
Dict2

xRef

ResultCheck
```
If you'd like to see it formatted some other way, please let me know :).

**Comment by vlastimilmaca** [comment]
Thank you very much for your time, I hope I fixed it all. Sorry my code is such a pain to get it right :).

**Comment by vlastimilmaca** [comment]
Just rebased to master (Travis didn't like how it was before, rebase seems to fixed it).

File: test/unit/annotation_spec.js
Comment (timvandermeij) [reply test/unit/annotation_spec.js]: Yes, I think that's even better. Let's go for that.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [reply test/unit/annotation_spec.js]: Looks really good this way!

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: Shouldn't this be `ReplyText`, just like in the test below?

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: I think you can even do `expect(data.inReplyTo).toEqual(annotationRef.toString());`.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: Same here regarding testing equality like in the test above.

File: test/unit/annotation_spec.js
Comment (timvandermeij) [inline test/unit/annotation_spec.js]: Just like in the tests above, we should check all fields instead of only the contents, i.e., contents, title, modification date and color (no need for the flags).

**Review by timvandermeij — APPROVED** [review]
I'm approving this for merging for when these final comments (and the one comment you indicated about moving the `TextAnnotation` unit tests below the `MarkupAnnotation` ones) have been addressed. Nice work!

File: test/unit/annotation_spec.js
Comment (vlastimilmaca) [reply test/unit/annotation_spec.js]: You are right, but I have to change few things to get it running. 

Test for `.toBeDefined` is nonsense in how it is now, because I set `annotationDict` into `IRT` property. In actual `annotations.js` code, there is test for `isRef`, and that ends up being null (Dict is not a Ref and null passes `toBeDefined `test … hence the "nonsense").

So instead of setting annotationDict I have to set annotationRef and add this section to make it pass:
```
 annotationDict.assignXref(xref);
 popupDict.assignXref(xref);
 replyDict.assignXref(xref);
```
Setting of `Dict `into properties where should be `Ref `seems to be used in other tests too and I think it is probably not correct, because it is not what is in real PDFs? I'm not sure … I don't understand this part of PDF world enough :).

**Comment by vlastimilmaca** [comment]
Should be all fixed. Thank you :).

**Review by timvandermeij — APPROVED** [review]

**Comment by timvandermeij** [comment]
Thank you for your contribution! I think the end result is a patch that is very well tested and maintainable.

**Comment by vlastimilmaca** [comment]
Nice. Thank you :).

**Pull request merged**
